tests/toranj/test-023-multicast-traffic.py

```diff
@@ -0,0 +1,203 @@
+#     notice, this list of conditions and the following disclaimer in the
+#     documentation and/or other materials provided with the distribution.
+#  3. Neither the name of the copyright holder nor the
+#     names of its contributors may be used to endorse or promote products
+#     derived from this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+#  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+
+import time
+import wpan
+from wpan import verify
+
+#-----------------------------------------------------------------------------------------------------------------------
+# Test description: Multicast traffic
+#
+# Network topology
+#
+#     r1 ---- r2 ---- r3 ---- r4
+#             |               |
+#             |               |
+#            fed             sed
+#
+# Test covers the following multicast traffic:
+#
+# - r2  =>> link-local all-nodes.   Expected to receive on [r1, r2, r3, fed].
+# - r3  =>> mesh-local all-nodes.   Expected to receive on [r1, r2, r3, r4, fed].
+# - r3  =>> link-local all-routers. Expected to receive on [r2, r3, r4].
+# - r3  =>> mesh-local all-routers. Expected to receive on all routers.
+# - r1  =>> link-local all-thread.  Expected to receive on [r1, r2].
+# - fed =>> mesh-local all-thread.  Expected to receive on all nodes.
+# - r1  =>> specific address (on r2 and sed). Expected to receive on [r2, sed].
+#
+
+test_name = __file__[:-3] if __file__.endswith('.py') else __file__
+print '-' * 120
+print 'Starting \'{}\''.format(test_name)
+
+#-----------------------------------------------------------------------------------------------------------------------
+# Utility functions
+
+def send_mcast(src_node, src_addr, mcast_addr, recving_node, msg_len=30):
+    """
+    Send a multicast message with given `len` from `src_node` using `src_addr` to the multicast address `mcast_addr`.
+    Verify that the message is received on all nodes in `recving_node` list.
+    """
+    sender = src_node.prepare_tx(src_addr, mcast_addr, msg_len, mcast_hops=5)
+    recvers = [node.prepare_rx(sender) for node in recving_node]
+
+    wpan.Node.perform_async_tx_rx()
+
+    verify(sender.was_successful)
+    for recvr in recvers:
+        verify(recvr.was_successful)
```

Comment:
Should we validate that messages were not received on devices that should not receive them?

Reply:
I was considering this and wanted this as part of this test. There is currently no easy way to add this to the code (we need a new type "Node.Receiver" which is bound to a local port on a specific interface and listens in background and collects all received packet. The problem is how to get such a "receiver" to stop. We currently do stop a "receiver" when we receive all expected packets. Need to think about this more)

I have checked that we don't receive on other devices by adding them to list and then seeing that test does fail... :)